### PR TITLE
Improve perf for Index based span indexers

### PR DESCRIPTION
**Pull request by AndyAyersMS**
First, evaluate the actual index before invoking the underlying int indexer,
so that the jit doesn't think the span is address taken.

Second, improve the jit's ability to remove some redundant comparisions like
those that arise in computing the actual index.

### CODE REVIEW

**Comment by AndyAyersMS** [comment]
@briansull PTAL
cc @dotnet/jit-contrib @tarekgh 

jit-diffs shows impact mainly to the new `Index` based operations on spans:
```
Total bytes of diff: -1099 (0.00% of base)
    diff is an improvement.
Top file improvements by size (bytes):
        -873 : System.Private.CoreLib.dasm (-0.02% of base)
         -46 : System.Data.Common.dasm (0.00% of base)
         -32 : System.Private.Xml.dasm (0.00% of base)
         -26 : Microsoft.CodeAnalysis.dasm (0.00% of base)
         -18 : Microsoft.VisualBasic.dasm (-0.01% of base)
17 total files with size differences (17 improved, 0 regressed), 112 unchanged.
Top method improvements by size (bytes):
        -234 (-28.23% of base) : System.Private.CoreLib.dasm - ReadOnlySpan`1:get_Item(struct):struct:this (5 methods)
        -234 (-28.23% of base) : System.Private.CoreLib.dasm - Span`1:get_Item(struct):struct:this (5 methods)
        -118 (-34.20% of base) : System.Private.CoreLib.dasm - ReadOnlySpan`1:get_Item(struct):byref:this (5 methods)
        -102 (-27.72% of base) : System.Private.CoreLib.dasm - Span`1:get_Item(struct):byref:this (5 methods)
         -80 (-31.87% of base) : System.Private.CoreLib.dasm - EventPipePayloadDecoder:ReadUnalignedGuid(byref):struct
Top method improvements by size (percentage):
        -118 (-34.20% of base) : System.Private.CoreLib.dasm - ReadOnlySpan`1:get_Item(struct):byref:this (5 methods)
         -80 (-31.87% of base) : System.Private.CoreLib.dasm - EventPipePayloadDecoder:ReadUnalignedGuid(byref):struct
        -234 (-28.23% of base) : System.Private.CoreLib.dasm - ReadOnlySpan`1:get_Item(struct):struct:this (5 methods)
        -234 (-28.23% of base) : System.Private.CoreLib.dasm - Span`1:get_Item(struct):struct:this (5 methods)
        -102 (-27.72% of base) : System.Private.CoreLib.dasm - Span`1:get_Item(struct):byref:this (5 methods)
43 total methods with size differences (43 improved, 0 regressed), 124016 unchanged.
```
Performance on a simple benchmark (see dotnet/corefx#33376) improved from ~80ms to ~57ms, matching "equivalent" indexer using a discrete int and bool. (For reference, the int indexer is around 35ms).

**Comment by AndyAyersMS** [comment]
[OSX failure](https://ci.dot.net/job/dotnet_coreclr/job/master/job/x64_checked_osx10.12_innerloop_flow_prtest/10360/) is likely unrelated. Will see if it repros....

@dotnet-bot retest OSX10.12 x64 Checked Innerloop Build and Test

**Comment by tarekgh** [comment]
The Span changes looks good to me. I'll others review the jit changes as I am not familiar with.

File: src/jit/assertionprop.cpp [hunk 1]
```diff
@@ -3029,19 +3064,66 @@ GenTree* Compiler::optAssertionProp_RelOp(ASSERT_VALARG_TP assertions, GenTree*
...
+            tree->ChangeOperConst(GT_CNS_INT);
+            tree->gtIntCon.gtIconVal = 1;
+        }
+
+        newTree = fgMorphTree(tree);
+        DISPTREE(newTree);
+        return optAssertionProp_Update(newTree, tree, stmt);
+    }
+
+    // Else if we have an equality check involving a local, look for
```
Comment (briansull) [inline src/jit/assertionprop.cpp]: At this point the comment should read:
`Else check if we have an equality check involving a local`

Or you could move this comment down to line 3123

**Pull request merged**